test-framework/junit5/src/main/java/io/quarkus/test/junit/NativeTestExtension.java

```diff
@@ -112,26 +82,117 @@ private void ensureNoInjectAnnotationIsUsed(Class<?> testClass) {
+    private ExtensionState ensureStarted(ExtensionContext extensionContext) {
+        Class<?> testClass = extensionContext.getRequiredTestClass();
+        ensureNoInjectAnnotationIsUsed(testClass);
+
+        ExtensionContext root = extensionContext.getRoot();
+        ExtensionContext.Store store = root.getStore(ExtensionContext.Namespace.GLOBAL);
+        ExtensionState state = store.get(ExtensionState.class.getName(), ExtensionState.class);
+        TestProfile annotation = testClass.getAnnotation(TestProfile.class);
+        Class<? extends QuarkusTestProfile> selectedProfile = null;
+        if (annotation != null) {
+            selectedProfile = annotation.value();
+        }
+        boolean wrongProfile = !Objects.equals(selectedProfile, quarkusTestProfile);
+        if ((state == null && !failedBoot) || wrongProfile) {
+            if (wrongProfile) {
+                if (state != null) {
+                    try {
+                        state.close();
+                    } catch (Throwable throwable) {
+                        throwable.printStackTrace();
+                    }
+                }
             }
-            ClassInfo testClassInfo = instance.target().asClass();
-            if (testClassInfo.classAnnotation(DotName.createSimple(TestProfile.class.getName())) != null) {
-                throw new JUnitException(
-                        "@TestProfile is not supported in NativeImageTest tests. Offending class is " + testClassInfo.name());
+            PropertyTestUtil.setLogFileProperty();
+            try {
+                state = doNativeStart(extensionContext, selectedProfile);
+                store.put(ExtensionState.class.getName(), state);
+
+            } catch (Throwable e) {
+                failedBoot = true;
+                firstException = e;
             }
         }
+        return state;
+    }
+
+    private ExtensionState doNativeStart(ExtensionContext context, Class<? extends QuarkusTestProfile> profile)
+            throws Throwable {
+        quarkusTestProfile = profile;
+        TestResourceManager testResourceManager = null;
+        try {
+            Class<?> requiredTestClass = context.getRequiredTestClass();
+
+            Map<String, String> sysPropRestore = new HashMap<>();
+            sysPropRestore.put(ProfileManager.QUARKUS_TEST_PROFILE_PROP,
+                    System.getProperty(ProfileManager.QUARKUS_TEST_PROFILE_PROP));
+
+            QuarkusTestProfile profileInstance = null;
+            final Map<String, String> additional = new HashMap<>();
+            if (profile != null) {
+                profileInstance = profile.newInstance();
+                additional.putAll(profileInstance.getConfigOverrides());
+                final Set<Class<?>> enabledAlternatives = profileInstance.getEnabledAlternatives();
+                if (!enabledAlternatives.isEmpty()) {
+                    additional.put("quarkus.arc.selected-alternatives", enabledAlternatives.stream()
```

Comment:
I don't think this can be supported as it is a build time property only

Reply:
Good point, but I think the current solution is not that bad: If `enabledAlternatives` is a non-empty set containing anything different than what the test app was compiled with, the test app will fail at boot. I think that's better than ignoring silently or warning. The user should do a conscious decision what to do in such a case. If he wants to test with the alternatives in JVM mode and ignore them in the native mode, he can create subclass of his QuarkusTestProfile for the native mode in which he returns an empty set of alternatives.

Do you think warn+ignore would be better?

Reply:
You are probably right, Let's keep it for now and see.